File: python/pysearch/searchsettings.py
# -*- coding: utf-8 -*-
###############################################################################
#
# searchsettings.py
#
# class SearchSettings: encapsulates search settings
#
###############################################################################
import re
from filetypes import FileType
from searchexception import SearchException


class SearchSettings(object):
    """a class to encapsulate search settings for a particular search session"""

    _extension_set_names = set('''in_extensions out_extensions
                                  in_archiveextensions
                                  out_archiveextensions'''.split())
# TODO: move filetypes to own group
    _pattern_set_names = set('''in_dirpatterns out_dirpatterns
                                in_filepatterns out_filepatterns
                                in_filetypes out_filetypes
                                in_archivefilepatterns out_archivefilepatterns
                                in_linesafterpatterns out_linesafterpatterns
                                in_linesbeforepatterns out_linesbeforepatterns
                                linesaftertopatterns linesafteruntilpatterns
                                searchpatterns'''.split())
    _props_with_defaults = {
        'archivesonly': False,
        'debug': False,
        'firstmatch': False,
        'excludehidden': True,
        'linesafter': 0,
        'linesbefore': 0,
        'listdirs': False,
        'listfiles': False,
        'listlines': False,
        'maxlinelength': 150,
        'multilinesearch': False,
        'printresults': True,
        'printusage': False,
        'printversion': False,
        'recursive': True,
        'searcharchives': False,
        'uniquelines': False,
        'verbose': False,
    }

    def __init__(self):
        self.startpath = ''
        for name in self._extension_set_names:
            self.__dict__[name] = set()
        for name in self._pattern_set_names:
            self.__dict__[name] = set()
        self.__dict__.update(self._props_with_defaults)

    def add_exts(self, exts, ext_set_name):
        if type(exts) is list:
            self.__dict__[ext_set_name] = self.__dict__[ext_set_name].union(exts)
        elif type(exts) is str:
            ext_set = set([ext for ext in exts.split(',') if ext])
            self.__dict__[ext_set_name] = self.__dict__[ext_set_name].union(ext_set)

    def add_patterns(self, patterns, pattern_set_name):
        assert pattern_set_name in self._pattern_set_names
        compile_flag = re.S | re.U
        if type(patterns) is list:
            pattern_set = set([re.compile(p, compile_flag) for p in patterns])
            self.__dict__[pattern_set_name] = self.__dict__[pattern_set_name].union(pattern_set)
        elif type(patterns) is str:
            self.__dict__[pattern_set_name].add(re.compile(patterns, compile_flag))
        else:
            raise SearchException('patterns is an unknown type')

    def add_filetypes(self, filetypes, filetype_set_name):
        # filetype_set = set()
        if type(filetypes) is list:
            filetype_set = set([FileType.from_name(ft) for ft in filetypes])
        elif type(filetypes) is str:
            filetype_set = set([FileType.from_name(ft) for ft in filetypes.split(',') if ft])
        else:
            raise SearchException('filetypes is an unknown type')
        self.__dict__[filetype_set_name] = self.__dict__[filetype_set_name].union(filetype_set)

    def set_property(self, name, val):
        self.__dict__[name] = val
        # some trues trigger others
        if type(val) is bool and val:
            if name == 'archivesonly':
                self.searcharchives = True
            elif name == 'debug':
                self.verbose = True

    def set_properties(self, propdict):
        for p in propdict:
            self.set_property(p, propdict[p])

    def __str__(self):
        all_props = {'startpath'} | self._extension_set_names | \
                    self._pattern_set_names | set(self._props_with_defaults.keys())
        print_dict = {}
        s = '{0}('.format(self.__class__.__name__)
        for p in sorted(all_props):
            val = self.__dict__[p]
            if type(val) == set:
                if p in self._pattern_set_names:
                    print_dict[p] = str([x.pattern for x in val])
                else:
                    print_dict[p] = str(list(val))
            elif type(val) is str:
                if val:
                    print_dict[p] = '"{0}"'.format(val)
                else:
                    print_dict[p] = '""'
            else:
                print_dict[p] = '{0!s}'.format(val)
        next_elem = 0
        for p in sorted(print_dict.keys()):
            if next_elem:
                s += ', '
            s += '{0}: {1}'.format(p, print_dict[p])
            next_elem += 1
        s += ')'
        return s

File: ruby/rbsearch/searchsettings.rb
################################################################################
#
# searchsettings.rb
#
# class SearchSettings: encapsulates search settings
#
################################################################################

class SearchSettings

  attr_accessor :archivesonly
  attr_accessor :debug
  attr_accessor :excludehidden
  attr_accessor :firstmatch
  attr_accessor :in_archiveextensions
  attr_accessor :in_archivefilepatterns
  attr_accessor :in_dirpatterns
  attr_accessor :in_extensions
  attr_accessor :in_filepatterns
  attr_accessor :in_linesafterpatterns
  attr_accessor :in_linesbeforepatterns
  attr_accessor :linesafter
  attr_accessor :linesaftertopatterns
  attr_accessor :linesafteruntilpatterns
  attr_accessor :linesbefore
  attr_accessor :listdirs
  attr_accessor :listfiles
  attr_accessor :listlines
  attr_accessor :maxlinelength
  attr_accessor :multilinesearch
  attr_accessor :out_archiveextensions
  attr_accessor :out_archivefilepatterns
  attr_accessor :out_dirpatterns
  attr_accessor :out_extensions
  attr_accessor :out_filepatterns
  attr_accessor :out_linesafterpatterns
  attr_accessor :out_linesbeforepatterns
  attr_accessor :printresults
  attr_accessor :printusage
  attr_accessor :printversion
  attr_accessor :recursive
  attr_accessor :searcharchives
  attr_accessor :searchpatterns
  attr_accessor :startpath
  attr_accessor :uniquelines
  attr_accessor :verbose

  def initialize
    @archivesonly = false
    @debug = false
    @excludehidden = true
    @firstmatch = false
    @linesafter = 0
    @linesbefore = 0
    @listdirs = false
    @listfiles = false
    @listlines = false
    @maxlinelength = 150
    @multilinesearch = false
    @printresults = true
    @printusage = false
    @printversion = false
    @recursive = true
    @searcharchives = false
    @startpath = nil
    @uniquelines = false
    @verbose = false

    @in_archiveextensions = []
    @in_archivefilepatterns = []
    @in_dirpatterns = []
    @in_extensions = []
    @in_filepatterns = []
    @in_linesafterpatterns = []
    @in_linesbeforepatterns = []
    @linesaftertopatterns = []
    @linesafteruntilpatterns = []
    @out_archiveextensions = []
    @out_archivefilepatterns = []
    @out_dirpatterns = []
    @out_extensions = []
    @out_filepatterns = []
    @out_linesafterpatterns = []
    @out_linesbeforepatterns = []
    @searchpatterns = []
  end

  def add_comma_delimited_exts(exts, ext_set)
    exts.split(",").each do |x|
        ext_set.push(x)
    end
  end

  def add_pattern(pattern, pattern_set)
    pattern_set.push(Regexp.new(pattern))
  end

  def list_to_s(name, lst)
    "#{name}=[\"#{lst.join("\", \"")}\"]"
  end

  def to_s
    s = 'SearchSettings('
    s += "archivesonly: #{@archivesonly}"
    s += ", debug: #{@debug}"
    s += ", firstmatch: #{@firstmatch}"
    s += ", excludehidden: #{@excludehidden}"
    s += ", " + list_to_s("in_archiveextensions", @in_archiveextensions)
    s += ", " + list_to_s("in_archivefilepatterns", @in_archivefilepatterns)
    s += ", " + list_to_s("in_dirpatterns", @in_dirpatterns)
    s += ", " + list_to_s("in_extensions", @in_extensions)
    s += ", " + list_to_s("in_filepatterns", @in_filepatterns)
    s += ", " + list_to_s("in_linesafterpatterns", @in_linesafterpatterns)
    s += ", " + list_to_s("in_linesbeforepatterns", @in_linesbeforepatterns)
    s += ", linesafter: #{@linesafter}"
    s += ", " + list_to_s("linesaftertopatterns", @linesaftertopatterns)
    s += ", " + list_to_s("linesafteruntilpatterns", @linesafteruntilpatterns)
    s += ", linesbefore: #{@linesbefore}"
    s += ", listdirs: #{@listdirs}"
    s += ", listfiles: #{@listfiles}"
    s += ", listlines: #{@listlines}"
    s += ", maxlinelength: #{@maxlinelength}"
    s += ", multilinesearch: #{@multilinesearch}"
    s += ", " + list_to_s("out_archiveextensions", @out_archiveextensions)
    s += ", " + list_to_s("out_archivefilepatterns", @out_archivefilepatterns)
    s += ", " + list_to_s("out_dirpatterns", @out_dirpatterns)
    s += ", " + list_to_s("out_extensions", @out_extensions)
    s += ", " + list_to_s("out_filepatterns", @out_filepatterns)
    s += ", " + list_to_s("out_linesafterpatterns", @out_linesafterpatterns)
    s += ", " + list_to_s("out_linesbeforepatterns", @out_linesbeforepatterns)
    s += ", printresults: #{@printresults}"
    s += ", printusage: #{@printusage}"
    s += ", printversion: #{@printversion}"
    s += ", recursive: #{@recursive}"
    s += ", searcharchives: #{@searcharchives}"
    s += ", " + list_to_s("searchpatterns", @searchpatterns)
    s += ", startpath: \"#{@startpath}\""
    s += ", uniquelines: #{@uniquelines}"
    s += ", verbose: #{@verbose}"
    s += ")"
    s
  end
end
